File: android/app/src/main/kotlin/com/splitfree/ui/components/Components.kt
package com.splitfree.ui.components

import androidx.compose.animation.animateColorAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.splitfree.core.Money
import com.splitfree.data.ParticipantEntity
import com.splitfree.ui.theme.Metrics
import com.splitfree.ui.theme.MoneyType
import com.splitfree.ui.theme.splitColors

/** A person's avatar: coloured initials, with an optional selection ring. */
@Composable
fun Avatar(
    participant: ParticipantEntity,
    size: Dp = Metrics.avatarMedium,
    showsRing: Boolean = false,
    dimmed: Boolean = false,
) {
    val colors = splitColors
    val background = colors.avatarColor(participant.colorIndex)
    Box(
        modifier = Modifier
            .size(size)
            .clip(CircleShape)
            .background(if (dimmed) background.copy(alpha = 0.35f) else background)
            .then(
                if (showsRing) Modifier.border(2.5.dp, colors.accent, CircleShape) else Modifier
            ),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = participant.initials,
            color = Color.White.copy(alpha = if (dimmed) 0.7f else 1f),
            fontSize = (size.value * 0.38f).sp,
            fontWeight = FontWeight.SemiBold,
        )
    }
}

/** Overlapping avatars with a "+N" cap: a group row's at-a-glance roster. */
@Composable
fun AvatarStack(
    participants: List<ParticipantEntity>,
    size: Dp = Metrics.avatarSmall,
    maxVisible: Int = 4,
) {
    val colors = splitColors
    val visible = participants.take(maxVisible)
    val overflow = (participants.size - maxVisible).coerceAtLeast(0)

    Row(horizontalArrangement = Arrangement.spacedBy((-size.value * 0.32f).dp)) {
        visible.forEach { person ->
            Box(modifier = Modifier.border(2.dp, colors.surface, CircleShape)) {
                Avatar(person, size)
            }
        }
        if (overflow > 0) {
            Box(
                modifier = Modifier
                    .size(size)
                    .border(2.dp, colors.surface, CircleShape)
                    .clip(CircleShape)
                    .background(colors.surfaceSunken),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    "+$overflow",
                    color = colors.secondaryText,
                    fontSize = (size.value * 0.34f).sp,
                    fontWeight = FontWeight.SemiBold,
                )
            }
        }
    }
}

/** The app's standard card. One radius, one border, used everywhere. */
@Composable
fun SplitCard(
    modifier: Modifier = Modifier,
    padding: PaddingValues = PaddingValues(Metrics.cardPadding),
    onClick: (() -> Unit)? = null,
    content: @Composable ColumnScope.() -> Unit,
) {
    val colors = splitColors
    Surface(
        modifier = modifier
            .fillMaxWidth()
            .then(if (onClick != null) Modifier.clickable(onClick = onClick) else Modifier),
        shape = RoundedCornerShape(Metrics.cardRadius),
        color = colors.surface,
        border = androidx.compose.foundation.BorderStroke(0.5.dp, colors.separator),
    ) {
        Column(modifier = Modifier.padding(padding), content = content)
    }
}

/** The rounded square that carries an expense's category. */
@Composable
fun CategoryBadge(color: Color, icon: ImageVector, size: Dp = 44.dp) {
    Box(
        modifier = Modifier
            .size(size)
            .clip(RoundedCornerShape(size * 0.3f))
            .background(color.copy(alpha = 0.14f)),
        contentAlignment = Alignment.Center,
    ) {
        Icon(icon, contentDescription = null, tint = color, modifier = Modifier.size(size * 0.5f))
    }
}

/** A balance rendered with the right color and the right words. */
@Composable
fun BalanceAmount(
    minorUnits: Long,
    currencyCode: String,
    style: androidx.compose.ui.text.TextStyle = MoneyType.row,
    settledText: String = "settled up",
) {
    val colors = splitColors
    if (minorUnits == 0L) {
        Text(settledText, color = colors.neutral, style = MaterialThemeBodyMedium())
    } else {
        val color by animateColorAsState(colors.forBalance(minorUnits), label = "balance")
        Text(
            Money(kotlin.math.abs(minorUnits), currencyCode).formatted(),
            color = color,
            style = style,
        )
    }
}

@Composable
private fun MaterialThemeBodyMedium() = androidx.compose.material3.MaterialTheme.typography.bodyMedium

/** A selectable pill: split methods, filters, date ranges. */
@Composable
fun Chip(
    label: String,
    selected: Boolean,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
    tint: Color? = null,
    onClick: () -> Unit,
) {
    val colors = splitColors
    val accent = tint ?: colors.accent
    Surface(
        modifier = modifier.clickable(onClick = onClick),
        shape = CircleShape,
        color = if (selected) accent else colors.surfaceSunken,
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 13.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            if (icon != null) {
                Icon(
                    icon,
                    contentDescription = null,
                    tint = if (selected) Color.White else colors.secondaryText,
                    modifier = Modifier.size(14.dp),
                )
            }
            Text(
                label,
                color = if (selected) Color.White else colors.secondaryText,
                style = androidx.compose.material3.MaterialTheme.typography.bodyMedium,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}

/** An empty state with an optional call to action. */
@Composable
fun EmptyState(
    icon: ImageVector,
    title: String,
    message: String,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null,
) {
    val colors = splitColors
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 44.dp, horizontal = Metrics.screenPadding),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Box(
            modifier = Modifier
                .size(88.dp)
                .clip(CircleShape)
                .background(colors.accentSoft),
            contentAlignment = Alignment.Center,
        ) {
            Icon(icon, contentDescription = null, tint = colors.accent, modifier = Modifier.size(34.dp))
        }
        Text(
            title,
            style = androidx.compose.material3.MaterialTheme.typography.titleLarge,
            color = colors.primaryText,
            textAlign = TextAlign.Center,
        )
        Text(
            message,
            style = androidx.compose.material3.MaterialTheme.typography.bodyMedium,
            color = colors.secondaryText,
            textAlign = TextAlign.Center,
            modifier = Modifier.width(300.dp),
        )
        if (actionLabel != null && onAction != null) {
            Spacer(Modifier.height(2.dp))
            SecondaryButton(actionLabel, onClick = onAction)
        }
    }
}

/**
 * The filled, affirmative action on a screen. There is at most one per card:
 * two of these side by side is how a destructive choice gets made by accident.
 */
@Composable
fun PrimaryButton(
    label: String,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val colors = splitColors
    Surface(
        modifier = modifier.clickable(enabled = enabled, onClick = onClick),
        shape = CircleShape,
        color = if (enabled) colors.accent else colors.accent.copy(alpha = 0.35f),
    ) {
        Text(
            label,
            modifier = Modifier.padding(horizontal = 22.dp, vertical = 12.dp),
            style = androidx.compose.material3.MaterialTheme.typography.labelLarge,
            color = Color.White,
        )
    }
}

@Composable
fun SecondaryButton(label: String, modifier: Modifier = Modifier, onClick: () -> Unit) {
    val colors = splitColors
    Surface(
        modifier = modifier.clickable(onClick = onClick),
        shape = CircleShape,
        color = colors.accent.copy(alpha = 0.12f),
    ) {
        Text(
            label,
            modifier = Modifier.padding(horizontal = 14.dp, vertical = 9.dp),
            color = colors.accent,
            fontWeight = FontWeight.SemiBold,
            style = androidx.compose.material3.MaterialTheme.typography.bodyMedium,
        )
    }
}

/** A subtle badge for states worth calling out without alarming anyone. */
@Composable
fun InfoBanner(text: String, icon: ImageVector, tint: Color? = null) {
    val colors = splitColors
    val accent = tint ?: colors.categoryBlue
    Surface(
        shape = RoundedCornerShape(Metrics.chipRadius),
        color = accent.copy(alpha = 0.09f),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Row(
            modifier = Modifier.padding(12.dp),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            Icon(icon, contentDescription = null, tint = accent, modifier = Modifier.size(16.dp))
            Text(
                text,
                style = androidx.compose.material3.MaterialTheme.typography.bodySmall,
                color = colors.secondaryText,
            )
        }
    }
}

/**
 * The persistent "does this add up?" strip used by every screen that
 * distributes a total. Green when balanced, red when not.
 */
@Composable
fun RemainderBanner(remainder: Long, currencyCode: String) {
    val colors = splitColors
    val balanced = remainder == 0L
    val message = when {
        balanced -> "Everything adds up"
        remainder > 0 -> "${Money(remainder, currencyCode).formatted()} left to assign"
        else -> "${Money(-remainder, currencyCode).formatted()} too much assigned"
    }
    Surface(
        shape = RoundedCornerShape(Metrics.chipRadius),
        color = if (balanced) colors.positiveSoft else colors.negativeSoft,
        modifier = Modifier.fillMaxWidth(),
    ) {
        Row(
            modifier = Modifier.padding(12.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Icon(
                Icons.Filled.Check,
                contentDescription = null,
                tint = if (balanced) colors.positive else colors.negative,
                modifier = Modifier.size(16.dp),
            )
            Text(
                message,
                color = if (balanced) colors.positive else colors.negative,
                fontWeight = FontWeight.SemiBold,
                style = androidx.compose.material3.MaterialTheme.typography.bodyMedium,
            )
        }
    }
}

/** Section heading with an optional trailing action. */
@Composable
fun SectionHeader(
    title: String,
    subtitle: String? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val colors = splitColors
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                title,
                style = androidx.compose.material3.MaterialTheme.typography.titleMedium,
                color = colors.primaryText,
            )
            if (subtitle != null) {
                Text(
                    subtitle,
                    style = androidx.compose.material3.MaterialTheme.typography.bodySmall,
                    color = colors.secondaryText,
                )
            }
        }
        trailing?.invoke()
    }
}

/** Small uppercase label used above figures. */
@Composable
fun Overline(text: String, modifier: Modifier = Modifier) {
    Text(
        text.uppercase(),
        modifier = modifier,
        style = androidx.compose.material3.MaterialTheme.typography.labelSmall,
        color = splitColors.tertiaryText,
    )
}

File: SplitFree/DesignSystem/Components.swift
import SwiftUI

// MARK: - Avatars

/// A person's avatar: their photo if they have one, otherwise coloured initials.
struct AvatarView: View {
    var participant: Participant
    var size: CGFloat = Metrics.avatarMedium
    var showsRing: Bool = false

    private var backgroundColor: Color {
        Palette.avatarColor(participant.colorIndex)
    }

    var body: some View {
        ZStack {
            if let data = participant.avatarData, let image = UIImage(data: data) {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFill()
            } else {
                backgroundColor
                Text(participant.initials)
                    .font(.system(size: size * 0.38, weight: .semibold, design: .rounded))
                    .foregroundStyle(.white)
                    .minimumScaleFactor(0.6)
                    .padding(2)
            }
        }
        .frame(width: size, height: size)
        .clipShape(Circle())
        .overlay {
            if showsRing {
                Circle().strokeBorder(Palette.accent, lineWidth: 2.5)
            }
        }
        .accessibilityHidden(true)
    }
}

/// Overlapping avatars with a "+N" cap - the group row's at-a-glance roster.
struct AvatarStack: View {
    var participants: [Participant]
    var size: CGFloat = Metrics.avatarSmall
    var maxVisible: Int = 4

    private var visible: [Participant] { Array(participants.prefix(maxVisible)) }
    private var overflow: Int { max(0, participants.count - maxVisible) }

    var body: some View {
        HStack(spacing: -size * 0.32) {
            ForEach(visible) { participant in
                AvatarView(participant: participant, size: size)
                    .overlay(Circle().strokeBorder(Palette.surface, lineWidth: 2))
            }
            if overflow > 0 {
                ZStack {
                    Circle().fill(Palette.surfaceSunken)
                    Text("+\(overflow)")
                        .font(.system(size: size * 0.34, weight: .semibold, design: .rounded))
                        .foregroundStyle(Palette.secondaryText)
                }
                .frame(width: size, height: size)
                .overlay(Circle().strokeBorder(Palette.surface, lineWidth: 2))
            }
        }
        .accessibilityElement()
        .accessibilityLabel(
            participants.count == 1
                ? participants[0].fullName
                : String(localized: "^[\(participants.count) person](inflect: true)", comment: "Member count")
        )
    }
}

/// The rounded square that carries an expense's category.
struct CategoryBadge: View {
    var category: ExpenseCategory
    var size: CGFloat = 44

    var body: some View {
        RoundedRectangle(cornerRadius: size * 0.3, style: .continuous)
            .fill(category.color.opacity(0.14))
            .frame(width: size, height: size)
            .overlay {
                Image(systemName: category.symbol)
                    .font(.system(size: size * 0.42, weight: .semibold))
                    .foregroundStyle(category.color)
            }
            .accessibilityHidden(true)
    }
}

/// Group identity tile - cover photo if set, otherwise a tinted glyph.
struct GroupBadge: View {
    var group: SpendingGroup
    var size: CGFloat = 48

    var body: some View {
        Group {
            if let data = group.coverImageData, let image = UIImage(data: data) {
                Image(uiImage: image).resizable().scaledToFill()
            } else {
                LinearGradient(
                    colors: [group.tint.opacity(0.9), group.tint.opacity(0.6)],
                    startPoint: .topLeading,
                    endPoint: .bottomTrailing
                )
                .overlay {
                    Image(systemName: group.kind.symbol)
                        .font(.system(size: size * 0.4, weight: .semibold))
                        .foregroundStyle(.white)
                }
            }
        }
        .frame(width: size, height: size)
        .clipShape(RoundedRectangle(cornerRadius: size * 0.28, style: .continuous))
        .accessibilityHidden(true)
    }
}

// MARK: - Containers

/// The app's standard card. One radius, one shadow, used everywhere.
struct Card<Content: View>: View {
    var padding: CGFloat = Metrics.cardPadding
    @ViewBuilder var content: Content

    var body: some View {
        content
            .padding(padding)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Palette.surface, in: RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous)
                    .strokeBorder(Palette.separator.opacity(0.6), lineWidth: 0.5)
            )
            .shadow(color: .black.opacity(0.045), radius: 12, x: 0, y: 4)
    }
}

/// A titled section with optional trailing action.
struct SectionHeader<Trailing: View>: View {
    var title: String
    var subtitle: String?
    @ViewBuilder var trailing: Trailing

    init(_ title: String, subtitle: String? = nil, @ViewBuilder trailing: () -> Trailing) {
        self.title = title
        self.subtitle = subtitle
        self.trailing = trailing()
    }

    var body: some View {
        HStack(alignment: .firstTextBaseline) {
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(Typography.sectionTitle)
                    .foregroundStyle(Palette.primaryText)
                if let subtitle {
                    Text(subtitle)
                        .font(Typography.caption)
                        .foregroundStyle(Palette.secondaryText)
                }
            }
            Spacer(minLength: 8)
            trailing
        }
        .accessibilityAddTraits(.isHeader)
    }
}

extension SectionHeader where Trailing == EmptyView {
    init(_ title: String, subtitle: String? = nil) {
        self.init(title, subtitle: subtitle) { EmptyView() }
    }
}

// MARK: - Money

/// A balance rendered with the right colour and the right words.
struct BalanceLabel: View {
    var minorUnits: Int
    var currencyCode: String
    var font: Font = Typography.rowMoney
    /// Shows "you owe" / "owes you" above the figure.
    var showsCaption: Bool = false
    var settledText: String = String(localized: "settled up")

    private var color: Color { .forBalance(minorUnits) }

    private var caption: String {
        if minorUnits > 0 { return String(localized: "you get back") }
        if minorUnits < 0 { return String(localized: "you owe") }
        return settledText
    }

    var body: some View {
        VStack(alignment: .trailing, spacing: 1) {
            if showsCaption {
                Text(caption)
                    .font(Typography.overline)
                    .textCase(.uppercase)
                    .foregroundStyle(Palette.tertiaryText)
                    .tracking(0.4)
            }
            if minorUnits == 0 && !showsCaption {
                Text(settledText)
                    .font(Typography.rowSubtitle)
                    .foregroundStyle(Palette.neutral)
            } else {
                Text(Money(minorUnits: abs(minorUnits), currencyCode: currencyCode).formatted())
                    .moneyStyle(font, color: color)
            }
        }
        .accessibilityElement()
        .accessibilityLabel(accessibilityText)
    }

    private var accessibilityText: String {
        let amount = Money(minorUnits: abs(minorUnits), currencyCode: currencyCode).formatted()
        if minorUnits > 0 {
            return String(format: String(localized: "You get back %@", comment: "Accessibility"), amount)
        }
        if minorUnits < 0 {
            return String(format: String(localized: "You owe %@", comment: "Accessibility"), amount)
        }
        return settledText
    }
}

// MARK: - Controls

/// The app's filled call to action.
struct PrimaryButtonStyle: ButtonStyle {
    var tint: Color = Palette.accent
    var isProminent: Bool = true

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.body.weight(.semibold))
            .foregroundStyle(isProminent ? Color.white : tint)
            .frame(maxWidth: .infinity)
            .frame(height: 52)
            .background {
                RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous)
                    .fill(isProminent ? AnyShapeStyle(tint) : AnyShapeStyle(tint.opacity(0.12)))
            }
            .scaleEffect(configuration.isPressed ? 0.975 : 1)
            .opacity(configuration.isPressed ? 0.9 : 1)
            .animation(Motion.quick, value: configuration.isPressed)
    }
}

/// Compact bordered button used in headers and toolbars.
struct SecondaryButtonStyle: ButtonStyle {
    var tint: Color = Palette.accent

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.subheadline.weight(.semibold))
            .foregroundStyle(tint)
            .padding(.horizontal, 14)
            .padding(.vertical, 9)
            .background(tint.opacity(0.12), in: Capsule())
            .scaleEffect(configuration.isPressed ? 0.96 : 1)
            .animation(Motion.quick, value: configuration.isPressed)
    }
}

/// A selectable pill - split methods, category filters, date ranges.
struct ChipButton: View {
    var title: String
    var systemImage: String?
    var isSelected: Bool
    var tint: Color = Palette.accent
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 6) {
                if let systemImage {
                    Image(systemName: systemImage)
                        .font(.system(size: 12, weight: .semibold))
                }
                Text(title)
                    .font(.subheadline.weight(.medium))
                    .lineLimit(1)
            }
            .foregroundStyle(isSelected ? Color.white : Palette.secondaryText)
            .padding(.horizontal, 13)
            .padding(.vertical, 8)
            .background {
                Capsule().fill(isSelected ? AnyShapeStyle(tint) : AnyShapeStyle(Palette.surfaceSunken))
            }
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(isSelected ? [.isSelected, .isButton] : .isButton)
    }
}

/// A row that reads as tappable without pretending to be a button.
struct DisclosureRow<Leading: View, Content: View>: View {
    @ViewBuilder var leading: Leading
    @ViewBuilder var content: Content

    var body: some View {
        HStack(spacing: 12) {
            leading
            content
            Spacer(minLength: 4)
            Image(systemName: "chevron.right")
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(Palette.tertiaryText)
        }
        .contentShape(Rectangle())
    }
}

// MARK: - Empty & loading states

struct EmptyStateView: View {
    var symbol: String
    var title: String
    var message: String
    var actionTitle: String?
    var action: (() -> Void)?

    var body: some View {
        VStack(spacing: 14) {
            ZStack {
                Circle()
                    .fill(Palette.accentSoft)
                    .frame(width: 88, height: 88)
                Image(systemName: symbol)
                    .font(.system(size: 34, weight: .medium))
                    .foregroundStyle(Palette.accent)
            }
            .padding(.bottom, 2)

            Text(title)
                .font(.title3.weight(.semibold))
                .foregroundStyle(Palette.primaryText)
                .multilineTextAlignment(.center)

            Text(message)
                .font(.subheadline)
                .foregroundStyle(Palette.secondaryText)
                .multilineTextAlignment(.center)
                .frame(maxWidth: 300)

            if let actionTitle, let action {
                Button(actionTitle, action: action)
                    .buttonStyle(SecondaryButtonStyle())
                    .padding(.top, 4)
            }
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 44)
        .padding(.horizontal, Metrics.screenPadding)
    }
}

/// A subtle badge for states worth calling out without alarming anyone.
struct InfoBanner: View {
    var symbol: String
    var text: String
    var tint: Color = Palette.categoryBlue

    var body: some View {
        HStack(alignment: .top, spacing: 10) {
            Image(systemName: symbol)
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(tint)
                .padding(.top, 1)
            Text(text)
                .font(.footnote)
                .foregroundStyle(Palette.secondaryText)
                .fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
        }
        .padding(12)
        .background(tint.opacity(0.09), in: RoundedRectangle(cornerRadius: Metrics.chipRadius, style: .continuous))
    }
}

// MARK: - Modifiers

/// Set by any pushed screen that has its own primary action, so the global
/// floating "add expense" button gets out of the way instead of hovering over
/// a detail view's own buttons.
struct HidesFloatingActionKey: PreferenceKey {
    static let defaultValue = false
    static func reduce(value: inout Bool, nextValue: () -> Bool) {
        value = value || nextValue()
    }
}

extension View {
    /// Standard screen chrome: app background, edge padding, large title.
    func screenBackground() -> some View {
        self.background(Palette.background.ignoresSafeArea())
    }

    func hidesFloatingAction() -> some View {
        self.preference(key: HidesFloatingActionKey.self, value: true)
    }

    /// Makes a whole row tappable with a press highlight, without the button look.
    func tappableRow(action: @escaping () -> Void) -> some View {
        Button(action: action) { self }
            .buttonStyle(RowButtonStyle())
    }
}

struct RowButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .background(
                RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous)
                    .fill(Palette.primaryText.opacity(configuration.isPressed ? 0.05 : 0))
            )
            .animation(Motion.quick, value: configuration.isPressed)
    }
}
